### PR TITLE
feat(camelCase/capitalize/kebabCase/lowerCase/snakeCase/startCase): improve return types

### PR DESCRIPTION
**Pull request by D-Sketon**
Even though the type passes the test, I can't guarantee that it covers all boundary cases

### CODE REVIEW

**Comment by D-Sketon** [comment]
close ＃172

File: src/string/_internal/getWords.ts [hunk 1]
```diff
@@ -16,6 +16,81 @@
...
+  | 'R'
+  | 'S'
+  | 'T'
+  | 'U'
+  | 'V'
+  | 'W'
+  | 'X'
+  | 'Y'
+  | 'Z';
+type LowercaseLetter = UppercaseLetter extends `${infer U}` ? Lowercase<U> : never;
```
Comment (StyleShit) [inline src/string/_internal/getWords.ts:50]: why do you need the `extends` here?

```suggestion
type LowercaseLetter = Lowercase<UppercaseLetter>;
```

File: src/string/capitalize.ts [hunk 2]
```diff
@@ -13,3 +13,5 @@
 export const capitalize = <T extends string>(str: T): Capitalize<T> => {
   return (str.charAt(0).toUpperCase() + str.slice(1).toLowerCase()) as Capitalize<T>;
 };
+
+export type Capitalize<T extends string> = T extends `${infer F}${infer R}` ? `${Uppercase<F>}${Lowercase<R>}` : T;
```
Comment (StyleShit) [inline src/string/capitalize.ts:17]: isn't it built in the language?

File: src/string/_internal/getWords.ts [hunk 3]
```diff
@@ -16,6 +16,81 @@
...
+type Digit = '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9';
+
+type EmptyStringFilter<L extends string[], S extends string> = S extends '' ? L : [...L, S];
+
+export type CaseSplit<
+  T extends string,
+  B extends string = '',
+  S extends string = '',
+  L extends string[] = [],
+> = string extends T
```
Comment (StyleShit) [inline src/string/_internal/getWords.ts:60]: maybe give those params better names to make it more readable?

File: src/string/_internal/getWords.ts [hunk 4]
```diff
@@ -16,6 +16,81 @@
...
+  | 'W'
+  | 'X'
+  | 'Y'
+  | 'Z';
+type LowercaseLetter = UppercaseLetter extends `${infer U}` ? Lowercase<U> : never;
+type Digit = '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9';
+
+type EmptyStringFilter<L extends string[], S extends string> = S extends '' ? L : [...L, S];
+
+export type CaseSplit<
```
Comment (StyleShit) [inline src/string/_internal/getWords.ts:55]: is it possible to separate this huge type into a bunch of smaller ones?

File: src/string/camelCase.spec.ts [hunk 5]
```diff
@@ -1,42 +1,53 @@
...
+  it('should work with screaming camel case', async () => {
+    expect(camelCase('FOO_BAR')).toEqual('fooBar');
+  });
+
+  it('handles return type properly', async () => {
+    expectTypeOf(camelCase('camelCase')).toEqualTypeOf<'camelCase'>();
+    expectTypeOf(camelCase('some whitespace')).toEqualTypeOf<'someWhitespace'>();
+    expectTypeOf(camelCase('hyphen-text')).toEqualTypeOf<'hyphenText'>();
+    expectTypeOf(camelCase('HTTPRequest')).toEqualTypeOf<'httpRequest'>();
+    expectTypeOf(camelCase('    leading and trailing whitespace')).toEqualTypeOf<'leadingAndTrailingWhitespace'>();
```
Comment (StyleShit) [inline src/string/camelCase.spec.ts:46]:
```suggestion
    expectTypeOf(camelCase('    leading and trailing whitespace    ')).toEqualTypeOf<'leadingAndTrailingWhitespace'>();
```

File: src/string/_internal/getWords.ts [hunk 6]
```diff
@@ -16,6 +16,81 @@
  */
 const CASE_SPLIT_PATTERN = /[A-Z]?[a-z]+|[0-9]+|[A-Z]+(?![a-z])/g;
 
-export function getWords(str: string): string[] {
-  return Array.from(str.match(CASE_SPLIT_PATTERN) ?? []);
+export function getWords<T extends string>(str: T): CaseSplit<T> {
+  return Array.from(str.match(CASE_SPLIT_PATTERN) ?? []) as CaseSplit<T>;
```
Comment (StyleShit) [inline src/string/_internal/getWords.ts:20]: I think you can do this:

```suggestion
  return Array.from(str.match(CASE_SPLIT_PATTERN) ?? []) as any;
```

or :

```suggestion
  return Array.from(str.match(CASE_SPLIT_PATTERN) ?? []) as never;
```

to make it shorter and easier to replace (now you'll have a single place to change if you'll have a slightly different return type)

WDYT?

File: src/string/camelCase.ts [hunk 7]
```diff
@@ -16,15 +18,33 @@ import { getWords } from "./_internal/getWords.ts";
  * const convertedStr4 = camelCase('HTTPRequest') // returns 'httpRequest'
  */
 
-export function camelCase(str: string): string {
+export function camelCase<T extends string>(str: T): CamelCase<T> {
   const words = getWords(str);
 
   if (words.length === 0) {
-    return "";
+    return '' as CamelCase<T>;
```
Comment (StyleShit) [inline src/string/camelCase.ts:25]: same as above regarding the `as`, and also other places

File: src/string/kebabCase.spec.ts [hunk 8]
```diff
@@ -45,4 +45,19 @@ describe('kebabCase', () => {
   it('should work with capitalized words', async () => {
     expect(kebabCase('Foo Bar')).toEqual('foo-bar');
   });
+
+  it('handles return type properly', async () => {
+    expectTypeOf(kebabCase('camelCase')).toEqualTypeOf<'camel-case'>();
+    expectTypeOf(kebabCase('some whitespace')).toEqualTypeOf<'some-whitespace'>();
+    expectTypeOf(kebabCase('hyphen-text')).toEqualTypeOf<'hyphen-text'>();
+    expectTypeOf(kebabCase('HTTPRequest')).toEqualTypeOf<'http-request'>();
+    expectTypeOf(kebabCase('    leading and trailing whitespace')).toEqualTypeOf<'leading-and-trailing-whitespace'>();
```
Comment (StyleShit) [inline src/string/kebabCase.spec.ts:54]:
```suggestion
    expectTypeOf(kebabCase('    leading and trailing whitespace    ')).toEqualTypeOf<'leading-and-trailing-whitespace'>();
```

File: src/string/lowerCase.spec.ts [hunk 9]
```diff
@@ -37,4 +37,17 @@ describe('lowerCase', () => {
   it('should work with screaming snake case', async () => {
     expect(lowerCase('FOO_BAR')).toEqual('foo bar');
   });
+
+  it('handles return type properly', async () => {
+    expectTypeOf(lowerCase('camelCase')).toEqualTypeOf<'camel case'>();
+    expectTypeOf(lowerCase('some whitespace')).toEqualTypeOf<'some whitespace'>();
+    expectTypeOf(lowerCase('hyphen-text')).toEqualTypeOf<'hyphen text'>();
+    expectTypeOf(lowerCase('HTTPRequest')).toEqualTypeOf<'http request'>();
+    expectTypeOf(lowerCase('    leading and trailing whitespace')).toEqualTypeOf<'leading and trailing whitespace'>();
```
Comment (StyleShit) [inline src/string/lowerCase.spec.ts:46]:
```suggestion
    expectTypeOf(lowerCase('    leading and trailing whitespace    ')).toEqualTypeOf<'leading and trailing whitespace'>();
```

**Review by StyleShit — CHANGES_REQUESTED** [review]

**Comment by raon0211** [comment]
Hello, thanks for your contribution to our library! While I understand the reason of providing exact types, I am not sure they are worth it, considering maintainability and readability of complex types. Let me think a little bit about whether or not to include complex types to our library.

**Comment by raon0211** [comment]
Hello, thanks for your contribution. 

Since we value [simplicity of the implementation](https://github.com/toss/es-toolkit/blob/main/.github/CONTRIBUTING.md#12-simplicity), we decided not to merge complex types in our repository.

However, other types of contributions such as adding new functions would be really helpful for us. Looking forward to your next contribution.
